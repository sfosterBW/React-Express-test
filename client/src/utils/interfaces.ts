export interface Fruit {
  id: string
  name: string
  best: boolean
  description?: string
}

export type NewFruit = Omit<Fruit, 'id'>
