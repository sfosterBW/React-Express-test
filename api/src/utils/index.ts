import { IFruit, NewFruit } from './types'

/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment*/

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

export const toNewFruit = (object: any): NewFruit => {
  const newFruit: NewFruit = {
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return newFruit
}

const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

export const toFruit = (object: any): IFruit => {
  const fruit: IFruit = {
    ...object,
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return fruit
}
