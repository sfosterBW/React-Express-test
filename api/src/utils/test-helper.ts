import Fruit from '../models/fruit'

const initialFruits = [
  {
    name: 'apple',
    best: false,
    description: "They keep doctors away"
  },
  {
    name: 'orange',
    best: true
  }
]

const fruitsInDb = async () => {
  const fruits = await Fruit.find({})
  return fruits
}

const nonExistingId = async () => {
  const fruit = new Fruit({
    name: 'willdelete',
    best: false
  })

  await fruit.save()
  await fruit.remove()

  return fruit.id
}

export default { initialFruits, fruitsInDb, nonExistingId }
