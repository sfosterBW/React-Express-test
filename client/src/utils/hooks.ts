import { ChangeEvent, useState } from 'react'

/* eslint-disable  @typescript-eslint/no-explicit-any */

interface Field<T> {
  label: string;
  name: string;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  type: string;
  value: T;
}

const useField = <T>(
  label: string,
  name: string,
  type: string,
  initValue: T
): [Field<T>, () => void] => {

  const [value, setValue] = useState(initValue)

  const onChange = (event: ChangeEvent<HTMLInputElement>): void => {
    if (type === 'checkbox' && typeof value === 'boolean') {
      setValue(event.target.checked as any)
    } else {
      setValue(event.target.value as any)
    }
  }

  const reset = (resetValue: T = initValue): void => {
    setValue(resetValue)
  }

  return [{
    onChange,
    label,
    name,
    type,
    value
  }, reset]
}

export default useField
