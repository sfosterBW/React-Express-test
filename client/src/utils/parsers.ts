import { Fruit } from './interfaces'

/* eslint-disable  @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`)
  }

  return best
}

export const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

const parseArray = (array: any): Array<unknown> => {
  if (!array || !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${String(array)}`)
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  const fruit: Fruit = {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }

  if (object.description) {
    fruit.description = parseDescription(object.description)
  }

  return fruit
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
