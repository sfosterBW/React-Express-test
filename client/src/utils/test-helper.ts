import { Fruit } from './interfaces'

export const emptyList: Fruit[] = []

export const fruit: Fruit = {
  id: "1",
  name:"apple",
  best:false,
  description: "I don't like doctors"
}

export const list: Fruit[] = [
  {
    id: "1",
    name: "apple",
    best: false,
    description: "I don't like doctors"
  },
  {
    id: "2",
    name: "banana",
    best: true
  }
]

export default { emptyList, fruit, list }
