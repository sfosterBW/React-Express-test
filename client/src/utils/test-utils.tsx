import React, { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from "redux-mock-store"
import thunk from "redux-thunk"

import { initState as initAlertState } from '../Reducers/alertReducer'
import { initState as initFruitState } from '../Reducers/fruitReducer'
import { initState as initModalState } from '../Reducers/modalReducer'
import { RootState } from '../Reducers/store'

export const rootInitialState = {
  alert: initAlertState,
  fruit: initFruitState,
  modal: initModalState
};

export const renderWithProviders = (
  ui: ReactNode,
  initialState: RootState = rootInitialState
) => {
  const store = configureStore([thunk])(initialState)
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    mockStore: store,
    getActionsTypes: () => store.getActions().map(a => a.type)
  }
}
