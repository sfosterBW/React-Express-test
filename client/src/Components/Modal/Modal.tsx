import React, { FC } from 'react'
import { RootState } from '../../Reducers/store'
import { useSelector, useDispatch } from 'react-redux'
import { toggleModal } from '../../Reducers/modalReducer'
import { RootActions } from '../../Reducers/store'
import { Fruit } from '../../utils/interfaces'
import EditForm from '../Form/EditForm'
import NewForm from '../Form/NewForm'
import styles from './Modal.module.css'

const Modal: FC<{fruit: Fruit | undefined}> = ({ fruit }) => {
  const dispatch = useDispatch()

  const toggle = useSelector((state: RootState) => state.modal.toggle)

  if(!toggle || !fruit) {
    return null
  }

  return (
    <div>
      <section className={styles.wrapper}>
        <div className={styles.modal}>
          {!fruit && <NewForm />}
          {fruit && <EditForm fruit={fruit} title="Edit fruit" />}
          <button
            className={styles.close}
            data-testid="close"
            name="close"
            onClick={(): RootActions => dispatch(toggleModal(!toggle))}
          >
            &times;
          </button>
        </div>
      </section>
    </div>
  )
}

export default Modal
