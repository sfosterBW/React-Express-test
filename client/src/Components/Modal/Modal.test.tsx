import React from 'react'
import { renderWithProviders, rootInitialState } from '../../utils/test-utils'
import { fruit } from '../../utils/test-helper'

import Modal from './Modal'

const component = <Modal fruit={fruit} />

describe('the modal component', () => {
  it('functions as expected', () => {
    const { getByTestId } = renderWithProviders(component, {
      ...rootInitialState,
      modal: { ...rootInitialState.modal, toggle: true }
    })
    expect(getByTestId('edit-form-title')).toBeTruthy()
  })

  it('renders the same as last time', () => {
    const { container } = renderWithProviders(component)
    expect(container).toMatchSnapshot()
  })
})
