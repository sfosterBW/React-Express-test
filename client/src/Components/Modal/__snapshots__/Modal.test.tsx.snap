// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the modal component renders the same as last time 1`] = `<div />`;
